br/pkg/lightning/restore/restore.go

```diff
@@ -1937,11 +1950,22 @@ func (rc *Controller) preCheckRequirements(ctx context.Context) error {
 
 	// We still need to sample source data even if this task has existed, because we need to judge whether the
 	// source is in order as row key to decide how to sort local data.
-	source, err := rc.estimateSourceData(ctx)
+	estimatedSizeResult, err := rc.preInfoGetter.EstimateSourceDataSize(ctx)
 	if err != nil {
 		return common.ErrCheckDataSource.Wrap(err).GenWithStackByArgs()
 	}
-	if rc.isLocalBackend() {
+	estimatedDataSizeToGenerate := estimatedSizeResult.SizeWithIndex
+
+	// Do not import with too large concurrency because these data may be all unsorted.
+	if estimatedSizeResult.HasUnsortedBigTables {
+		if rc.cfg.App.TableConcurrency > rc.cfg.App.IndexConcurrency {
+			rc.cfg.App.TableConcurrency = rc.cfg.App.IndexConcurrency
+		}
+	}
+	if rc.status != nil {
+		rc.status.TotalFileSize.Store(estimatedSizeResult.SizeWithoutIndex)
+	}
```

Comment:
where does this part come from? new added?

Reply:
It is originally in `br/pkg/lightning/restore/check_info.go` L445-L453